transformer_engine/pytorch/csrc/extensions/attention.cu

```diff
@@ -1419,3 +1419,599 @@ at::Tensor fa_prepare_bwd(at::Tensor q, at::Tensor k, at::Tensor v) {
+                                const at::Tensor &cu_seqlens,
+                                const std::string &first_half,
+                                const std::string &second_half) {
+  if (first_half == "add" && second_half == "none") {
+    thd_grad_correction_helper<dtype, AddFunctor<dtype>, EmptyFunctor, 0>(
+      grad, grad_per_step, cu_seqlens);
+  } else if (first_half == "copy" && second_half == "none") {
+    thd_grad_correction_helper<dtype, CopyFunctor, EmptyFunctor, 0>(
+      grad, grad_per_step, cu_seqlens);
+  } else if (first_half == "none" && second_half == "add") {
+    thd_grad_correction_helper<dtype, EmptyFunctor, AddFunctor<dtype>, 1>(
+      grad, grad_per_step, cu_seqlens);
+  } else if (first_half == "none" && second_half == "copy") {
+    thd_grad_correction_helper<dtype, EmptyFunctor, CopyFunctor, 1>(
+      grad, grad_per_step, cu_seqlens);
+  } else if (first_half == "add" && second_half == "copy") {
+    thd_grad_correction_helper<dtype, AddFunctor<dtype>, CopyFunctor, 2>(
+      grad, grad_per_step, cu_seqlens);
+  } else if (first_half == "copy" && second_half == "add") {
+    thd_grad_correction_helper<dtype, CopyFunctor, AddFunctor<dtype>, 2>(
+      grad, grad_per_step, cu_seqlens);
+  } else {
+    NVTE_ERROR("Unsupported Functor of first half and second_half\n");
+  }
+}
+
+void thd_grad_correction(at::Tensor &grad,
+                         const at::Tensor &grad_per_step,
+                         const at::Tensor &cu_seqlens,
+                         const std::string &first_half,
+                         const std::string &second_half) {
+  if (grad.scalar_type() == at::ScalarType::Half) {
+    thd_grad_dispatcher<at::Half>(grad, grad_per_step, cu_seqlens, first_half, second_half);
+  } else if (grad.scalar_type() == at::ScalarType::BFloat16) {
+    thd_grad_dispatcher<at::BFloat16>(grad, grad_per_step, cu_seqlens, first_half, second_half);
+  } else if (grad.scalar_type() == at::ScalarType::Float) {
+    thd_grad_dispatcher<float>(grad, grad_per_step, cu_seqlens, first_half, second_half);
+  } else {
+    NVTE_ERROR("Unsupported dtype of grad\n");
+  }
+}
+
+/***************************************************************************************************
+ * Support THD format for Context Parallel: Generate partitioned indices for input tokens
+ **************************************************************************************************/
+
+__global__ void thd_partition_indices_kernel(int *output,
+                                             int *cu_seqlens,
+                                             int batch,
+                                             int total_tokens,
+                                             int world_size,
+                                             int rank) {
+  extern __shared__ int cu_seqlens_s[];
+  for (int i = threadIdx.x; i <= batch; i += blockDim.x) {
+    cu_seqlens_s[i] = cu_seqlens[i] / world_size;
```

Comment:
assert check if each sequence length is divisible by CP*2 at here? we can remove the check after you have more general implementation for thd format.

Reply:
Added assert check.